lib/CGI.pm

```diff
@@ -622,16 +622,8 @@ sub init {
 	  last METHOD;
       }
 
-      # If method is GET, HEAD or DELETE, fetch the query from
-      # the environment.
-      if ($is_xforms || $meth=~/^(GET|HEAD|DELETE)$/) {
-          $query_string = $self->_get_query_string_from_env;
-         $self->param($meth . 'DATA', $self->param('XForms:Model'))
-             if $is_xforms;
-	      last METHOD;
-      }
-
-      if ($meth eq 'POST' || $meth eq 'PUT') {
+      # First, let's try to get data from request body
+      if ($meth=~/^(GET|HEAD|DELETE|POST|PUT)$/) {
```

Comment:
Yes, I wanted to be as generic as possible : let's try to find parameters, whatever they are (request body, query string), whatever the request method is (GET, HEAD, DELETE, POST, PUT).
All possible cases are then handled.